pkg/protocol/http/http.go

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright The Dragonfly Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package http
+
+import (
+	"net/http"
+
+	"github.com/dragonflyoss/Dragonfly/pkg/errortypes"
+	"github.com/dragonflyoss/Dragonfly/pkg/protocol"
+)
+
+var (
+	// DefaultClientOpt is default implementation of ClientOpt.
+	DefaultClientOpt = &ClientOpt{
+		transport: http.DefaultTransport,
+	}
+
+	// DefaultClient is default implementation of Client.
+	DefaultClient = &Client{
+		client:    &http.Client{Transport: DefaultClientOpt.transport},
+		transport: DefaultClientOpt.transport,
+	}
+)
+
+const (
+	// http protocol name
+	ProtocolName = "http"
+)
+
+func init() {
+	protocol.RegisterProtocol(ProtocolName, &ClientBuilder{})
+}
+
+// ClientOpt is the argument of NewProtocolClient.
+type ClientOpt struct {
+	transport http.RoundTripper
+}
+
+// Client is an implementation of protocol.Client for http protocol.
+type Client struct {
+	client    *http.Client
+	transport http.RoundTripper
+}
+
+func (cli *Client) GetResource(url string, md protocol.Metadata) protocol.Resource {
+	var (
+		hd *Headers
+	)
+
+	if md != nil {
+		h, ok := md.(*Headers)
+		if ok {
+			hd = h
+		}
+	}
+
+	if hd == nil {
+		hd = NewHTTPMetaData().(*Headers)
```

Comment:
Why must create it? I think, it can be `nil`.

Reply:
Allow it.